### ISSUE TITLE
Filtering the rows on a checkbox selection can lead to 2 rows being checked.

### ISSUE THREAD

**Issue by yehudag** [swimlane/ngx-datatable#824]
**I'm submitting a ...**  (check one with "x")
```
[ x] bug report => search github for a similar issue or PR before submitting
[ ] feature request
[ ] support request => Please do not submit support request here, post on Stackoverflow or Gitter
```

**Current behavior**

With checkbox selection enabled I select 1 row. On the select event I filter the rows. I end up with 2 rows being checked.

**Expected behavior**

Only 1 row should be checked.

**Reproduction of the problem**

[http://plnkr.co/edit/QpG2e1?p=preview](http://plnkr.co/edit/QpG2e1?p=preview)
Run the plunker and select the second row.

**What is the motivation / use case for changing the behavior?**

In my example plunker I used Bagel Shops so I will use to illustrate the issue. \
The plunker app displays a list of bagel shops that the user can visit on their trip to NYC.  In order to generate a tour route the user selects a few shops, that are near each other, they would like to visit. On the first selection I filter the rows to only display other shops that are nearby (filtered by zipcode). 

I would like to be able to restrict what a user can select based on the user's first selection. Currently I am filtering the rows on the first selection.

**Please tell us about your environment:**

Windows 7, VS Code, yarn, 
* **Table version:** 0.8.x

9.3.0
* **Angular version:** 2.0.x

4.2.2
* **Browser:** [Chrome 59 | Firefox 54]

* **Language:** [ TypeScript 2.3]

**Comment by yehudag** [comment]
I was able to workaround this issue by calling `ChangeDetctorRef.detectChanges()` before updating the rows, though I don't know what this says about the issue

**Comment by sarora2073** [comment]
This doesn't appear to be a problem with the package. 

If you replace your onSelect with the following, everything works fine for the normal use case:
    this.selected = []; // clear this.selected (which holds state) and push currently selected
    this.selected.push(...selected);    
    
However, if I add your code (re-copied below) to the above 2 lines, the problem occurs...at first blush this appears to be an async processing issue..your code modifies this.rows which is simultaneously being used by the checkbox handling code to determine which row to check. You could easily prove this theory by running your code with a setInterval delay. 

    if (selected.length > 0) {
      this.rows = this.bagelShops.filter(bagelShop => bagelShop.zipcode === selected[0].zipcode);
    } else {
      this.rows = this.bagelShops;
    }

**Comment by yehudag** [comment]
I agree that selection works for the normal case; I only encountered this issue when filtering the rows on the selection event. 

Regarding your second point:  since `ChangeDetctorRef.detectChanges` will wait until the change detection cycle has completed calling `detectChanges` before I update the rows accomplishes the same thing as setTimeout (I am assuming you meant setTimeout vs. setInterval).  Waiting for `detectChanges` ensures that the checkbox handling completes before I update the rows.

**Comment by sarora2073** [comment]
yes, setTimeout is what I meant. Glad you were able to resolve this in your code. 

From looking through the list of issues on this forum, there are open items that attempt to address how data is exposed in general (what should be immutable etc.), so it appears there is recognition of a broader problem. Not sure if you want to consider this one "closed"?

**Comment by yehudag** [comment]
I think it's worth keeping this issue open at least until #827 is resolved. If someone else runs into this issue they'll be able to find it. It's also a concrete example of how the issues mentioned #827 affects real use cases.
